crates/sui-core/src/congestion_tracker.rs

```diff
@@ -0,0 +1,331 @@
+                        .input_shared_objects()
+                        .into_iter()
+                        .filter_map(|object| match object {
+                            InputSharedObject::Mutate((id, _, _)) => Some(id),
+                            InputSharedObject::Cancelled(_, _)
+                            | InputSharedObject::ReadOnly(_)
+                            | InputSharedObject::ReadDeleted(_, _)
+                            | InputSharedObject::MutateDeleted(_, _) => None,
+                        })
+                        .collect::<Vec<_>>(),
+                ));
+            }
+        }
+
+        self.process_per_checkpoint_events(
+            checkpoint.timestamp_ms,
+            &congestion_events,
+            &cleared_events,
+        );
+    }
+
+    /// For all the mutable shared inputs, get the highest minimum clearing price (if any exists)
+    /// and the lowest maximum cancelled price.
+    pub fn get_suggested_gas_prices(&self, transaction: &TransactionData) -> Option<u64> {
+        self.get_suggested_gas_price_for_objects(
+            transaction
+                .shared_input_objects()
+                .into_iter()
+                .filter(|id| id.mutable)
+                .map(|id| id.id),
+        )
+    }
+}
+
+impl CongestionTracker {
+    fn process_per_checkpoint_events(
+        &self,
+        now: CheckpointTimestamp,
+        congestion_events: &[(u64, Vec<ObjectID>)],
+        cleared_events: &[(u64, Vec<ObjectID>)],
+    ) {
+        let congestion_info_map =
+            self.compute_per_checkpoint_congestion_info(now, congestion_events, cleared_events);
+        self.process_checkpoint_congestion(congestion_info_map);
+    }
+
+    fn get_suggested_gas_price_for_objects(
+        &self,
+        objects: impl Iterator<Item = ObjectID>,
+    ) -> Option<u64> {
+        let mut clearing_price = None;
+        for object_id in objects {
+            if let Some(info) = self.get_congestion_info(object_id) {
+                let clearing_price_for_object = match info
+                    .last_success_time
+                    .cmp(&Some(info.last_cancellation_time))
+                {
+                    std::cmp::Ordering::Greater => {
+                        // there were no cancellations in the most recent checkpoint,
+                        // so the object is probably not congested any more
+                        None
+                    }
+                    std::cmp::Ordering::Less => {
+                        // there were no successes in the most recent checkpoint. This should be a rare case,
+                        // but we know we will have to bid at least as much as the highest cancelled price.
+                        Some(info.highest_cancelled_gas_price)
+                    }
+                    std::cmp::Ordering::Equal => {
+                        // there were both successes and cancellations.
+                        info.lowest_executed_gas_price
```

Comment:
Can highest_cancelled_gas_price > lowest_executed_gas_price? If so, I wonder if this should be max(lowest_executed_gas_price, highest_cancelled_gas_price)

Reply:
In this case the last success and last cancel time indicate that the events are from the same checkpoint. So no, I don't believe the highest cancelled could be higher than the lowest executed

Reply:
Actually, I guess it could, because a transaction could have been cancelled because it was also using some other even more expensive object. But in this case lowest executed is still probably the right thing.